sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2.scala

```diff
@@ -117,20 +118,43 @@ object DataWritingSparkTask extends Logging {
       writeTask: DataWriterFactory[InternalRow],
       context: TaskContext,
       iter: Iterator[InternalRow]): WriterCommitMessage = {
-    val dataWriter = writeTask.createDataWriter(context.partitionId(), context.attemptNumber())
+    val stageId = context.stageId()
+    val partId = context.partitionId()
+    val attemptId = context.attemptNumber()
+    val dataWriter = writeTask.createDataWriter(partId, attemptId)
 
     // write the data and commit this writer.
     Utils.tryWithSafeFinallyAndFailureCallbacks(block = {
       iter.foreach(dataWriter.write)
-      logInfo(s"Writer for partition ${context.partitionId()} is committing.")
-      val msg = dataWriter.commit()
-      logInfo(s"Writer for partition ${context.partitionId()} committed.")
+
+      val msg = if (writeTask.useCommitCoordinator) {
+        val coordinator = SparkEnv.get.outputCommitCoordinator
```

Comment:
I'm not sure why we need this. In the implementation of `DataWriter.commit`, users can still call `SparkEnv.get. outputCommitCoordinator`. User can even use their own commit coordinator which is based on zookeeper or something.

I think the current API is flexible enough to: 1) not use commit coordinator 2) use Spark built-in commit coordinator 3) use customer commit coordinator.

Reply:
The API is flexible. The problem is that it defaults to no coordination, which cause correctness bugs.

The safe option is to coordinate commits by default. If an implementation doesn't change the default, then it at least won't duplicate task outputs in job commit. Worst case is that it takes a little longer for committers that don't need coordination. On the other hand, not making this the default will cause some writers to work most of the time, but duplicate data in some cases.

What do you think is the down side to adding this?

Reply:
Yea it makes sense to use a commit coordinator by default, but I think we need to carefully design the API to introduce the concept of commit coordinator, just a `boolean useCommitCoordinator()` seems not enough. We also need to update the documentation of the write APIs, to clearly specify in which phase the commit coordinator is involved and how it works.

Reply:
What do you have in mind to "introduce the concept"?

I'm happy to add more docs. Do you want me to add them to this PR or in a follow-up? Are you targeting this for 2.3.0?

Reply:
Since we have a workaround(call coordinator in `DataWriter.commit`), I don't think this should block the 2.3 release, but we can definitely get this in branch 2.3 if there is no breaking change on the public APIs.

And I won't treat it as a correctness bug. The default no-coordinator behavior is well documented with the current APIs, see the classdoc of `DataWriter`. We never guarantee that for an RDD partition, only one task can commit successfully.

> What do you have in mind to "introduce the concept"?

I never thought about it before, I'll think about it these days.

Reply:
Let me know if you want me to change this PR.

I'd like to see this go into 2.3.0 if there's still time. Just because it is documented doesn't mean it isn't a choice that severely limits the utility of DataSourceV2. I'd rather not support work-arounds for the life of 2.3.0.

Reply:
I updated the `DataSourceWriter` docs for this change.

Reply:
bq. We never guarantee that for an RDD partition, only one task can commit successfully

There's at-least once though, right? And then the Job Commit (which is implicitly at-most-once) is expected to handle the situation wherein 1+ task may have committed, and should resolve it so that the output of only one task is added.

One thing which I think would be good is for the spark docs to somewhere (scaladoc? markdown) to precisely write down its requirements of a committer. For the WiP paper on the new S3A committers, [I've tried to do this across MR & Spark](https://github.com/steveloughran/zero-rename-committer/blob/master/tex/a_zero_rename_committer.tex#L1993)

1. Complete: you get the output of all committed tasks
2. Exclusive: you only get the output of committed tasks
3. (Consistent: produces right output even if store is inconsistent)
4. Concurrent: >1 task may commit simultaneously
5. Abortable: if you abort a task, no output is visible
6. Continuity of correctness: after a job is committed,  no partitioned task may suddenly add its work to the output.

Not required: if there's a partition and a 2nd task attempt is committed, the output of either one of those attempts must be committed, but the specifics of which one is left open.

* Hadoop MR v1 meets 1-6 on HDFS, fails on 3 against raw S3
* The Direct Parquet committer fails to meet requirements (2, 5 & probably 6)
* The Hadoop MR v2 committer fails on 2, because if a task attempt commit fails partway through, some of its output may be in the dest dir. Both Spark and MR assume that this situation never occurs. Really, committers should be able to say "Doesn't support retry on task commit failure", or better. 

Regarding this patch,

1. how often do you actually expect people to be doing their own commit co-ordinator? 
1. What's the likelihood that they will get it right?

As we can see, the number of people who can correctly implement a committer is << than those who have shipped one; I don't see a commit coordinator being any different. It's good to offer the flexibility, but important to have the default being the one which everyone else uses and which is generally trusted.

Reply:
> As we can see, the number of people who can correctly implement a committer is << than those who have shipped one

Totally agree. Great quote.

From Wenchen's comments, I think we're in agreement that the default should be to use the commit coordinator. We just need to figure out how to get it in and how to document what it guarantees.